CARDS-1761: Adapt React Date / Time pickers

veronikaslc opened the pull request:
Remake of [old PR.](https://github.com/data-team-uhn/cards/pull/1147) Fixes:

[CARDS-1761](https://phenotips.atlassian.net/browse/CARDS-1761)
[CARDS-1979](https://phenotips.atlassian.net/browse/CARDS-1979)
[CARDS-1926](https://phenotips.atlassian.net/browse/CARDS-1926)

Affected components:
- Filters
- Date/Time questions
- Downtime Warning Configuration
- Patient identification screen

[CARDS-1761]: https://phenotips.atlassian.net/browse/CARDS-1761?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-1979]: https://phenotips.atlassian.net/browse/CARDS-1979?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-1926]: https://phenotips.atlassian.net/browse/CARDS-1926?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
CARDS-1761: Adapt React Date / Time pickers
- Remaned DateQuestionUtils to TimeQuestionUtils;
- Adapted DateTimePicker to the Question components;
- Refactored DateQuestionMonth and DateQuestionFull components into one DateQuestion component;
- Adapted DatePicker in the DowntimeWarningConfiguration and PatientIdentification components;
- Adapted DateTimePicker in the DateFilter component;

veronikaslc committed:
CARDS-1979: Add ability to CLEAR the value of a date/time input in the date/time picker

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Moved DateTimeUtilities to commons package

veronikaslc committed:
[misc] Use checkPropTypes wrapper

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Fixed naming after rebase

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Deleted DropdownsDatePicker

acrowthe commented on modules/commons/src/main/frontend/src/components/DateTimeUtilities.jsx:
```diff
@@ -105,16 +108,15 @@ export default class DateQuestionUtilities {
+    let truncateTo = 'second';
     for (let [formatSpecifier, targetPrecision] of Object.entries(truncate)) {
-      if (toFormat.indexOf(formatSpecifier) < 0) {
+      if (toFormat.indexOf(formatSpecifier) < 0 && ("H" != formatSpecifier || toFormat.indexOf("h") < 0)) {
```
This logic feels like it could use an explanation:
```suggestion
      // Check if the current specifier is missing.
      // If the current specifier is 'H', check if both 'H' and 'h' are missing
      if (toFormat.indexOf(formatSpecifier) < 0 && ("H" != formatSpecifier || toFormat.indexOf("h") < 0)) {
```

acrowthe commented on modules/commons/src/main/frontend/src/components/DateTimeUtilities.jsx:
```diff
@@ -247,6 +238,40 @@ export default class DateQuestionUtilities {
+  }
+
+  static formatIsHourMinuteSeconds(dateFormat) {
+    return typeof(dateFormat) === "string" && dateFormat.toLowerCase() === "HH:mm:ss";
```
```suggestion
    return typeof(dateFormat) === "string" && dateFormat.toLowerCase() === "hh:mm:ss";
```

acrowthe commented:
All these were tested using the Date Formats Test form

Bug: When there is an invalid date in a date that has more than one section, the error message sometimes shows `null`:
![image](https://github.com/user-attachments/assets/cebc7a9d-fa9c-4863-8e2e-d3090fc4be13)
Reproduction:
- Start from an empty picker that has at least 2 sections eg. a year-month picker, a year-month range, a year down to second picker
- Partially fill out the date picker, eg. fill out month but not year or year but not month, then lose focus on the picker
  - Observe that the error message is correct
- Fill out the picker completely
- Clear one part of the picker, eg remove day or year, keeping the other part(s) complete
  - Observe that the error message shows `null`

Bug:
Error messages which show the entered date in the message do not update with the current value of the input:
![image](https://github.com/user-attachments/assets/d6f4cc35-2573-4f65-8ef5-633115eadbb6)
Reproduction:
- Start with a completely blank question that has at least two sections  eg. a year-month picker, a year-month range, a year down to second picker.
- Into the year section, type in 2025.
- Lose focus on the question by clicking elsewhere on the page
- Observe that the error message reads the date as 0002-<placeholders>, not 2025-<placeholders>

Bug:
Filling out one field of a range input hides any errors that should be shown for the other field
![image](https://github.com/user-attachments/assets/9f210324-5cb0-4802-a839-571cc5fbe4ac)
Reproduction:
- Fill out start or end of a range with an error, eg. fill out start year but not month.
- Lose focus on the question, observe that the error is correct
- Fill out the the other end/start of the range, leaving the errored field as is eg. fill out end year and month.
- Lose focus. Observe that the error message is now gone, despite the field still being invalid

Bug:
There are no triggered errors if dates outside of the valid range are entered by typing in the date, rather than trying to use the date picker pop up
![image](https://github.com/user-attachments/assets/581ef86a-1b83-4bad-a678-b23c8e2c56cc)

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Display date error messages in edit and view mode

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers

Add comments

Co-authored-by: acrowthe <acrowthe3388@gmail.com>

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers

fix date format case

Co-authored-by: acrowthe <acrowthe3388@gmail.com>

sdumitriu commented on modules/commons/src/main/frontend/src/components/DateTimeUtilities.jsx:
```diff
@@ -124,7 +128,7 @@ export default class DateQuestionUtilities {
   static dateToFormattedString(date, textFieldType) {
     return (!date?.isValid) ? "" :
-    textFieldType === "date" ? date.toFormat("yyyy-MM-dd") : date.toFormat("yyyy-MM-dd\'T\'HH:mm");
+    textFieldType === "date" ? date.toFormat(this.VIEW_DATE_FORMAT) : date.toFormat("yyyy-MM-dd\'T\'HH:mm");
```
Why the change from `yyyy-MM-dd` to `yyyy/MM/dd`? If the year is first, most of the time the separator is `-`. When the separator is `/`, it is usually because the actual date format is `dd/MM/yyyy` but written in reverse for right-to-left locales.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,279 @@
+  if (questionDefinition.dataType === "date") {
+    let dateType = DateTimeUtilities.getDateType(questionDefinition.dateFormat);
+    if ( [DateTimeUtilities.FULL_DATE_TYPE, DateTimeUtilities.DATETIME_TYPE, DateTimeUtilities.MONTH_DATE_TYPE].includes(dateType)) {
+      return [StyledDateQuestion, 60];
```
This has the same priority as `DateQuestionYear`, to avoid ambiguity it should have a higher value.

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/PatientIdentification.jsx:
```diff
@@ -324,8 +332,40 @@ function PatientIdentification(props) {
+                                 }
+                               },
+                               inputProps: {
+                                 placeholder: `${dateFormat}, for example ${DateTime.fromISO("1970-12-31").toFormat(dateFormat)}`
```
This is only taken into account when opening the date picker, most of the time it just displays the label (`Date of birth`).

sdumitriu reviewed (commented):
Weird behavior when switching between typing and picking dates:
- create a new Visit Information form
- type 2022 in the Time answer, press Tab
- answer gets red, error that the date is invalid (correct)
- click on the picker, select a valid date
- answer is changed with the selected date (correct), but the validation error still remains (wrong)
- click on the input, delete some of the fields, e.g. `2025-MM-DD 00:00`
- the validation error is not retriggered (wrong)
- this last step is actually an error even if the picker is not used, e.g. type in a complete date, then delete some of the parts, the error is not displayed (wrong)

Timezone and milliseconds don't play nice with the picker. If the format is `yyyy-MM-dd'T'HH:mm:ss.SSSZ`, then `.SSSZ` is displayed verbatim in the input. If the format is `yyyy-MM-dd'T'HH:mm:ssZ`, then it doesn't recognize that the last part is seconds+timezone and displays `ssZ` in the input.

Milliseconds are not supported at all, if the stored date does include milliseconds, then upon editing and saving that precision is lost, the field is always truncated to seconds.

I'm OK with the milliseconds not being displayed, pretending that the desired date format doesn't include `\\.S+` and `Z` and `x+`, but I would prefer if milliseconds are not lost when the date is not changed at all.

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Addressed comments

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
More consistent error validation on manual and picker date manipulation

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Added milliseconds to the default sling format as well as truncating

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Added answer instructions to the top

acrowthe commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -228,7 +228,6 @@ function ExportButton(props) {
     return (<LocalizationProvider dateAdapter={AdapterLuxon}>
               <DateTimePicker
                 label="Any date"
-                views={['year', 'month', 'day', 'hours', 'minutes']}
```
Feels like this should be clearable since empty date fields are valid in the export window and trying to manually empty the new date pickers is tedious

acrowthe commented on modules/downtime-warning-banner/src/main/frontend/src/DowntimeWarningConfiguration.jsx:
```diff
@@ -92,40 +126,10 @@ function DowntimeWarningConfiguration() {
-                placeholder={dateFormat.toLowerCase()}
-                value={fromDate || ""}
-              />
+              {getDateField("Start of maintenance", fromDate, setFromDate)}
```
The downtime configuration can not be saved if the date fields are blank. The was possible with the previous date picker.

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -26,9 +26,9 @@ import { withStyles } from 'tss-react/mui';
 import Answer, {VALUE_POS} from "./Answer";
 import AnswerComponentManager from "./AnswerComponentManager";
-import DateQuestionUtilities from "./DateQuestionUtilities";
+import DateTimeUtilities from "../components/DateTimeUtilities";
```
The Computed date and time questions no longer seem to work - The computed questions in the `Computed Field Test` questionnaire never get filled out.
<img width="328" height="287" alt="image" src="https://github.com/user-attachments/assets/f5e97986-8b4a-476d-be59-42ea0596ddf7" />

acrowthe replied:
The latest changes make computed dates work but it's not perfect - the computed questions display a date in most cases but the output formatting is not as defined and in some cases no output is present.
<img width="335" height="300" alt="image" src="https://github.com/user-attachments/assets/1867efb3-a137-410d-b574-892c4f86cb80" />
<img width="412" height="287" alt="image" src="https://github.com/user-attachments/assets/39f82592-323e-42a5-9d3a-b4cbc12d6a2e" />
In view mode, the dates are formatted correctly and present even if edit mode is incorrect:
<img width="429" height="293" alt="image" src="https://github.com/user-attachments/assets/f1b31810-6b86-4084-a647-f10b921abad6" />

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,298 @@
+      }
+      // Determine if the end date is earlier than the start date
+      if (isRange && displayedDate && displayedEndDate && displayedEndDate < displayedDate) {
+        setRangeError(rangeErrorMessage);
```
This only shows up the second time an invalid range is entered if the date picker popup is used.
Testing:
- In runmode test, create a new Date Formats Test form
- In any of the interval fields without time (such as Month Year Interval), enter the start date as the current day/month/year
- On the end date, click the calendar icon to open the date picker, then choose the previous day/month. Observe that the error is not shown.
- On the end date, use the calendar icon to open the date picker and enter the day or month before that. Observe that the invalid range error is not shown

This error is delayed if editing the start date by date picker as well. This check works fine if the date format includes a time or if the day/month/year is typed in instead of entered using the date picker popup.

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionYear.jsx:
```diff
@@ -62,25 +62,25 @@ function DateQuestionYear(props) {
       maxAnswers={1}
       dataType="long"
-      isRange={(type === DateQuestionUtilities.INTERVAL_TYPE)}
+      isRange={(type === DateTimeUtilities.INTERVAL_TYPE)}
```
With changes to year questions, it feels like the interval question in the test questionnaire should be un-commented for testing. It seems to work fine, not sure why it was excluded originally:
https://github.com/data-team-uhn/cards/blob/b7b6bfab5296fc28f67c963d20340529c1cd277d/test-resources/src/main/resources/SLING-INF/content/Questionnaires/DateFormatsTest.xml#L208

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Addressed codereview comments

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Removed leftover code from using moment; Fixed saving empty date for DowntimeWarningConfiguration

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Removed duplicated methods from DateTimeUtilities
Moved non-utilities methods from DateTimeUtilities to DateQuestion
Optimised answerType calculation in ComputedQuestion

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Use DateTimeUtilities formatDateAnswer() method to display date string

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Do not strip the timezone for the date processing

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Correct meridiem/hour detection
